src/cargo/core/compiler/job_queue.rs

```diff
@@ -839,30 +844,24 @@ impl<'cfg> DrainState<'cfg> {
         }
 
         let time_elapsed = util::elapsed(cx.bcx.config.creation_time().elapsed());
-        if let Err(e) = self.timings.finished(cx, &error) {
-            if error.is_some() {
-                crate::display_error(&e, &mut cx.bcx.config.shell());
-            } else {
-                return Some(e);
-            }
+        if let Err(e) = self.timings.finished(cx, &errors.to_error()) {
+            self.handle_error(&mut cx.bcx.config.shell(), &mut errors, e);
         }
```

Comment:
In response to:

> I'm finding it a bit confusing to follow why above cases use `handle_error` but `self.timings.finished` and the `writeln!` below both use different custom logic. I think, however, that it should be fine to use `handle_error` in both cases?

Reply:
Yea, I think it should be fine.  It's been a while, but I think my thought process was that "if something has gone wrong, might as well exit right away".   The consequence is that you might end up with multiple errors displayed.